libs/langchain/langchain/vectorstores/google/generativeai/semantic_retriever.py

```diff
@@ -0,0 +1,170 @@
+                is None
+            ):
+                raise DoesNotExistsException(
+                    corpus_id=corpus_id, document_id=document_id
+                )
+
+        return cls(name=name)
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[Dict[str, Any]]] = None,
+        document_id: Optional[str] = None,
+    ) -> List[str]:
+        if self.name.document_id is None and document_id is None:
+            raise NotImplementedError(
+                "Adding texts to a corpus directly is not supported. "
+                "Please provide a document ID under the corpus first. "
+                "Then add the texts to the document."
+            )
+        if (
+            self.name.document_id is not None
+            and document_id is not None
+            and self.name.document_id != document_id
+        ):
+            raise NotImplementedError(
+                f"Parameter `document_id` {document_id} does not match the "
+                f"vector store's `document_id` {self.name.document_id}"
+            )
+        assert self.name.document_id or document_id is not None
+        new_document_id = self.name.document_id or document_id or ""
+
+        texts = list(texts)
+        if metadatas is None:
+            metadatas = [{} for _ in texts]
+        if len(texts) != len(metadatas):
+            raise ValueError(
+                f"metadatas's length {len(metadatas)} and "
+                f"texts's length {len(texts)} are mismatched"
+            )
+
+        chunks = genaix.batch_create_chunk(
+            corpus_id=self.name.corpus_id,
+            document_id=new_document_id,
+            texts=texts,
+            metadatas=metadatas,
+            client=self._client,
+        )
+
+        return [chunk.name for chunk in chunks if chunk.name]
+
+    def similarity_search(
+        self,
+        query: str,
+        k: int = 4,
+        filter: Optional[Dict[str, Any]] = None,
+    ) -> List[Tuple[str, float]]:
+        if self.name.is_corpus():
+            relevant_chunks = genaix.query_corpus(
+                corpus_id=self.name.corpus_id,
+                query=query,
+                k=k,
+                filter=filter,
+                client=self._client,
+            )
+        else:
+            assert self.name.is_document()
+            assert self.name.document_id is not None
+            relevant_chunks = genaix.query_document(
+                corpus_id=self.name.corpus_id,
+                document_id=self.name.document_id,
+                query=query,
+                k=k,
+                filter=filter,
+                client=self._client,
+            )
+
+        return [
+            (chunk.chunk.data.string_value, chunk.chunk_relevance_score)
+            for chunk in relevant_chunks
+        ]
+
+    def delete(self, ids: Optional[List[str]] = None, **kwargs: Any) -> Optional[bool]:
+        for id in ids or []:
```

Comment:
Future optimization: if these are chunks in the same document, they can be deleted in one BatchDeleteChunks call.